Taken from Oracle solaris-userland b83a793,
components/x11/xserver/xorg/patches/27-sparc-multisession.patch (xorg-server 21.1.24).

Solaris sparc specific patch for multi session setup

Not suitable for upstream

--- a/config/hal.c	2019-02-12 12:37:35.112189147 -0800
+++ b/config/hal.c	2019-02-12 12:40:26.404813665 -0800
@@ -1,7 +1,7 @@
 /*
  * Copyright © 2007 Daniel Stone
  * Copyright © 2007 Red Hat, Inc.
- * Copyright (c) 2013, 2025, Oracle and/or its affiliates.
+ * Copyright (c) 2013, 2025, Oracle and/or its affiliates.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
@@ -150,7 +150,7 @@
     char *ptr;
     char disppath[PATH_MAX];
 
-    if ((num_session_disp_dev == num_total_disp_dev) || !disp_dev_path[0])
+    if (((num_session_disp_dev & 0xFF) == num_total_disp_dev) || !disp_dev_path[0])
        return FALSE;
 
     if (lstat(path, &statbuf) == 0 &&
@@ -494,7 +494,7 @@
     }
 
 #if ((defined(__sparc__) || defined(__sparc)) && defined(SUNSOFT))
-    if ((num_session_disp_dev < num_total_disp_dev) &&
+    if ((num_session_disp_dev < (num_total_disp_dev & 0x0FF)) &&
                (!strcmp(name, "keyboard") || !strcmp(name, "mouse"))) {
        int i;
 
--- a/hw/xfree86/common/xf86AutoConfig.c	2019-02-12 12:40:47.029843557 -0800
+++ b/hw/xfree86/common/xf86AutoConfig.c	2019-02-12 12:43:58.523857933 -0800
@@ -2,6 +2,7 @@
  * Copyright 2003 by David H. Dawes.
  * Copyright 2003 by X-Oz Technologies.
  * All rights reserved.
+ * Copyright (c) 2013, 2025, Oracle and/or its affiliates.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
@@ -51,6 +52,11 @@
 #include <ctype.h>
 #endif
 
+#if (defined(__sparc__) || defined(__sparc))
+static Bool MultiSessionConfig (void);
+extern int num_total_disp_dev;
+#endif
+
 /* Sections for the default built-in configuration. */
 
 #define BUILTIN_DEVICE_NAME \
@@ -200,6 +206,11 @@
         return FALSE;
     }
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined (sun))
+    /* Do not do MultiSessionConfig() when invoked with -isolateDevice option */
+    if ((num_total_disp_dev & 0x1000)|| !MultiSessionConfig()) {
+#endif
+
     listPossibleVideoDrivers(&md);
 
     for (i = 0; i < md.nmatches; i++) {
@@ -231,6 +242,10 @@
         free(md.matches[i]);
     }
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined (sun))
+    }
+#endif
+
     xf86MsgVerb(X_DEFAULT, 0,
                 "Using default built-in configuration (%d lines)\n",
                 builtinLines);
@@ -468,3 +483,174 @@
 
     return ptr;
 }
+
+#if ((defined(__sparc__) || defined(__sparc)) && defined (sun))
+#include <libdevinfo.h>
+#include <xorg/xf86Pci.h>
+#include "pciaccess.h"
+#include <sys/utsname.h>
+
+#define MAX_DEV                        16
+
+#define IS_VGA(c) \
+       (((c) & 0x00ffff00) \
+       == ((PCI_CLASS_DISPLAY << 16) | (PCI_SUBCLASS_DISPLAY_VGA << 8)))
+
+
+#define BUILTIN_MULTI_SERVERFLAGS_OPT1 \
+       "\tOption\t\"DefaultServerLayout\"\t\"Xsession0\"\n"
+
+#define BUILTIN_MULTI_SECTION_POST \
+       "EndSection\n\n"
+
+#define BUILTIN_MULTI_SERVERFLAGS_OPT2 \
+       "\tOption\t\"AutoAddDevices\"\t\"true\"\n"
+
+#define BUILTIN_MULTI_SERVERFLAGS_SECTION \
+       "Section \"ServerFlags\"\n" \
+       BUILTIN_MULTI_SERVERFLAGS_OPT1 \
+       BUILTIN_MULTI_SERVERFLAGS_OPT2 \
+       BUILTIN_MULTI_SECTION_POST
+
+#define BUILTIN_MULTI_LAYOUT_SECTION_PRE \
+       "Section \"ServerLayout\"\n" \
+      "\tIdentifier\t\"Xsession%d\"\n"
+
+#define BUILTIN_MULTI_LAYOUT_SECTION_LINE \
+       "\tScreen\t\t\"Screen%d\"\n"
+
+#define BUILTIN_MULTI_SCREEN_SECTION_PRE \
+       "Section \"Screen\"\n" \
+       "\tIdentifier\t\"Screen%d\"\n"
+
+#define BUILTIN_MULTI_SCREEN_SECTION_LINE \
+       "\tDevice\t\t\"Card%d\"\n"
+
+#define BUILTIN_MULTI_DEVICE_SECTION_PRE \
+       "Section \"Device\"\n" \
+       "\tIdentifier\t\"Card%d\"\n"
+
+#define BUILTIN_MULTI_DEVICE_SECTION_LINE \
+       "\tBusID\t\t\"%s\"\n\tDriver\t\t\"%s\"\n"
+
+static Bool
+MultiSessionConfig (void)
+{
+    struct pci_device_iterator * iter;
+    struct pci_device * dev;
+    typedef struct {
+       char            busid[64];
+       char            dev_path[PATH_MAX];
+    } disp_dev_type;
+    disp_dev_type disp_dev[MAX_DEV];
+    int num = 0;
+    char buf[1024];
+    int i;
+    struct utsname sys;
+    struct sol_device_private {
+        struct pci_device  base;
+        const char * device_string;
+    };
+#define DEV_PATH(dev)    (((struct sol_device_private *) dev)->device_string)
+
+    if (uname(&sys) < 0) {
+       xf86Msg(X_ERROR, "Error in uname call\n");
+       return FALSE;
+    }
+
+    if (strcmp(sys.machine, "sun4v"))
+       return FALSE;
+
+    iter = pci_slot_match_iterator_create( NULL );
+
+    while ( (dev = pci_device_next( iter )) != NULL ) {
+       if (IS_VGA(dev->device_class)) {
+           if (num > MAX_DEV) {
+               xf86Msg(X_ERROR, "Too many display devices: %d\n", num);
+               return FALSE;
+           }
+
+           memset(&disp_dev[num], 0, sizeof (disp_dev_type));
+
+           snprintf(disp_dev[num].busid, sizeof(disp_dev[num].busid),
+               "PCI:%d@%d:%d:%d", dev->bus, dev->domain,
+               dev->dev, dev->func);
+
+           snprintf(disp_dev[num].dev_path, sizeof(disp_dev[num].dev_path),
+               "/devices%s", DEV_PATH(dev));
+
+           num++;
+       }
+    }
+
+    pci_iterator_destroy(iter);
+
+    /* Do not do multi-session configuration if dev number is 0 or 1 */
+    if (num <= 1)
+       return FALSE;
+
+    snprintf(buf, sizeof(buf), BUILTIN_MULTI_SERVERFLAGS_SECTION);
+    AppendToConfig(buf);
+
+    for (i = 0; i < num; i++) {
+       char drv[32];
+
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_LAYOUT_SECTION_PRE, i);
+       AppendToConfig(buf);
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_LAYOUT_SECTION_LINE, i);
+       AppendToConfig(buf);
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_SECTION_POST);
+       AppendToConfig(buf);
+
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_SCREEN_SECTION_PRE, i);
+       AppendToConfig(buf);
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_SCREEN_SECTION_LINE, i);
+       AppendToConfig(buf);
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_SECTION_POST);
+       AppendToConfig(buf);
+
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_DEVICE_SECTION_PRE, i);
+       AppendToConfig(buf);
+
+       drv[0] = 0;
+       if (disp_dev[i].dev_path[0] != '\0') {
+           int iret;
+           int fd;
+
+
+           fd = open(disp_dev[i].dev_path, O_RDONLY);
+           if (fd >= 0) {
+               struct vis_identifier visid;
+               const char *cp;
+
+               SYSCALL(iret = ioctl(fd, VIS_GETIDENTIFIER, &visid));
+               close (fd);
+
+               if (iret >= 0) {
+                   if (strcmp(visid.name, "SUNWtext") != 0) {
+                       for (cp = visid.name; (*cp != '\0') && isupper((uchar_t)*cp); cp++);
+                       /* find end of all uppercase vendor section */
+                       if ((cp != visid.name) && (*cp != '\0'))
+                           strlcpy (drv, cp, sizeof(drv));
+                   }
+               }
+           }
+        }
+
+       if (drv[0] == 0) {
+           xf86Msg(X_ERROR, "Can't find driver for session %d\n", i);
+           FreeConfig();
+           return FALSE;
+       }
+
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_DEVICE_SECTION_LINE,
+               disp_dev[i].busid, drv);
+       AppendToConfig(buf);
+       snprintf(buf, sizeof(buf), BUILTIN_MULTI_SECTION_POST);
+      AppendToConfig(buf);
+    }
+
+    return TRUE;
+}
+#endif
+
--- a/hw/xfree86/common/xf86pciBus.c	2019-02-12 12:44:13.372840348 -0800
+++ b/hw/xfree86/common/xf86pciBus.c	2019-02-12 12:45:25.611287613 -0800
@@ -130,6 +130,27 @@
     }
     free(iter);
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined (sun))
+    /*
+     * num_total_disp_dev does not reflect accurate number of display
+     * device when xf86IsolateDevice is set, redo it.
+     */
+    if (xf86IsolateDevice.domain != PCI_MATCH_ANY) {
+       num_total_disp_dev = 0x1000;
+
+       iter = pci_slot_match_iterator_create (NULL);
+       while ((info = pci_device_next(iter)) != NULL) {
+           if (PCIINFOCLASSES(info->device_class)) {
+               pci_device_probe(info);
+
+               if (IS_VGA(info->device_class))
+                   num_total_disp_dev++;
+           }
+       }
+    }
+    free(iter);
+#endif
+
     /* If we haven't found a primary device try a different heuristic */
     if (primaryBus.type == BUS_NONE && num) {
         for (i = 0; i < num; i++) {
@@ -1221,7 +1242,11 @@
 #if defined(__linux__) || defined(__NetBSD__)
         driverList[idx++] = "nouveau";
 #endif
+#ifdef sun
+        driverList[idx++] = "nvidia";
+#else
         driverList[idx++] = "nv";
+#endif
         break;
     }
     case 0x1106:
--- a/hw/xfree86/common/xf86Xinput.c	2019-02-12 12:45:50.385519021 -0800
+++ b/hw/xfree86/common/xf86Xinput.c	2019-02-12 12:46:53.201042630 -0800
@@ -45,9 +45,9 @@
  * the sale, use or other dealings in this Software without prior written
  * authorization from the copyright holder(s) and author(s).
  */
-+/*
-+ * Copyright (c) 2013, Oracle and/or its affiliates. All Rights Reserved.
-+ */
+/*
+ * Copyright (c) 2013, 2014, Oracle and/or its affiliates.
+ */
 
 #ifdef HAVE_XORG_CONFIG_H
 #include <xorg-config.h>
@@ -1567,7 +1567,7 @@
     }
 
 #if ((defined(__sparc__) || defined(__sparc)) && defined(sun))
-    if (num_session_disp_dev < num_total_disp_dev) {
+    if (num_session_disp_dev < (num_total_disp_dev & 0x0FFF)) {
        int i;
 
        for (i = 0; i < MAX_DEVICES; i++) {
--- a/hw/xfree86/common/xf86AutoConfig.c	2025-10-28 07:23:12.831417367 -0700
+++ b/hw/xfree86/common/xf86AutoConfig.c	2025-10-28 07:22:52.671402314 -0700
@@ -274,64 +274,6 @@
 #ifdef XSERVER_PLATFORM_BUS
     xf86PlatformMatchDriver(md);
 #endif
-#ifdef __sun
-    /* Check for driver type based on /dev/fb type and if valid, use
-       it instead of PCI bus probe results */
-    if (xf86Info.consoleFd >= 0) {
-        struct vis_identifier visid;
-        const char *cp;
-        int iret;
-
-        SYSCALL(iret = ioctl(xf86Info.consoleFd, VIS_GETIDENTIFIER, &visid));
-        if (iret < 0) {
-            int fbfd;
-
-            fbfd = open(xf86SolarisFbDev, O_RDONLY);
-            if (fbfd >= 0) {
-                SYSCALL(iret = ioctl(fbfd, VIS_GETIDENTIFIER, &visid));
-                close(fbfd);
-            }
-        }
-
-        if (iret < 0) {
-            xf86Msg(X_WARNING,
-                    "could not get frame buffer identifier from %s\n",
-                    xf86SolarisFbDev);
-        }
-        else {
-            xf86Msg(X_PROBED, "console driver: %s\n", visid.name);
-
-            /* Special case from before the general case was set */
-            if (strcmp(visid.name, "NVDAnvda") == 0) {
-                xf86AddMatchedDriver(md, "nvidia");
-            }
-
-            /* General case - split into vendor name (initial all-caps
-               prefix) & driver name (rest of the string). */
-            if (strcmp(visid.name, "SUNWtext") != 0) {
-                for (cp = visid.name; (*cp != '\0') && isupper(*cp); cp++) {
-                    /* find end of all uppercase vendor section */
-                }
-                if ((cp != visid.name) && (*cp != '\0')) {
-                    char *vendorName = xnfstrdup(visid.name);
-
-                    vendorName[cp - visid.name] = '\0';
-
-                    xf86AddMatchedDriver(md, vendorName);
-                    xf86AddMatchedDriver(md, cp);
-
-                    free(vendorName);
-                }
-            }
-        }
-    }
-#endif
-#ifdef __sparc__
-    const char *sbusDriver = sparcDriverName();
-
-    if (sbusDriver)
-        xf86AddMatchedDriver(md, sbusDriver);
-#endif
 #ifdef XSERVER_LIBPCIACCESS
     xf86PciMatchDriver(md);
 #endif
--- a/hw/xfree86/common/xf86Bus.c	2019-03-13 13:18:04.774729935 -0700
+++ b/hw/xfree86/common/xf86Bus.c	2019-03-13 13:18:16.025013289 -0700
@@ -247,9 +247,6 @@
 #ifdef XSERVER_LIBPCIACCESS
     xf86PciProbe();
 #endif
-#if (defined(__sparc__) || defined(__sparc)) && !defined(__OpenBSD__)
-    xf86SbusProbe();
-#endif
 #ifdef XSERVER_PLATFORM_BUS
     xf86platformPrimary();
 #endif
